### PR TITLE
Move choose from cython to backend

### PR DESCRIPTION
**Pull request by LukichevaPolina**
Move choose from cython to backend

### CODE REVIEW

File: dpnp/dpnp_iface_indexing.py [hunk 1]
```diff
@@ -81,35 +81,29 @@ def choose(x1, choices, out=None, mode='raise'):
...
-            pass
-        elif not isinstance(choices, list):
+    x1_desc = dpnp.get_dpnp_descriptor(x1)
+    
+    choices_list =[]
+    for choice in choices:
+        choices_list.append(dpnp.get_dpnp_descriptor(choice))
+
+    size = x1_desc.size
+    choices_size = choices_list[0].size
```
Comment (shssf) [inline dpnp/dpnp_iface_indexing.py:91]: `dpnp.get_dpnp_descriptor` can return `False` if no conversion possible. In this case, it should go to fallback

File: dpnp/dpnp_algo/dpnp_algo_indexing.pyx [hunk 2]
```diff
@@ -65,12 +66,27 @@ ctypedef void(*custom_indexing_6in_func_ptr_t)(void *, void * , void * , const s
 ctypedef void(*fptr_dpnp_nonzero_t)(const void * , void * , const size_t, const size_t * , const size_t , const size_t)
 
 
-cpdef utils.dpnp_descriptor dpnp_choose(object input, list choices):
-    cdef shape_type_c obj_shape = utils._object_to_tuple(len(input))
-    cdef utils.dpnp_descriptor res_array = utils_py.create_output_descriptor_py(obj_shape, choices[0].dtype, None)
+cpdef utils.dpnp_descriptor dpnp_choose(utils.dpnp_descriptor input, list choices1):
+    cdef shape_type_c input_shape = input.shape  
+
+    cdef vector[void * ] choices
```
Comment (shssf) [inline dpnp/dpnp_algo/dpnp_algo_indexing.pyx:72]:
```suggestion
    cdef vector[utils.dpnp_descriptor] choices
```

File: dpnp/backend/kernels/dpnp_krnl_indexing.cpp [hunk 3]
```diff
@@ -26,12 +26,46 @@
...
+void dpnp_choose_c(void* result1, void* array1_in, std::vector<void*> choices1, size_t size)
+{
+    if ((array1_in == nullptr) || (result1 == nullptr) || !choices1.size() || !size)
+    {
+        return;
+    }
+    DPNPC_ptr_adapter<_DataType1> input1_ptr(array1_in, size);
+    _DataType1* array_in = input1_ptr.get_ptr();
+
+    DPNPC_ptr_adapter<_DataType2*> choices_ptr(choices1.data(), choices1.size());
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_indexing.cpp:49]: Wrong usage.
In this code, potentially, the container itself could be copied. The main idea is to copy input memory from `device` to the kernel space. So, it looks like a loop needed here over input data stored in the container

File: dpnp/backend/kernels/dpnp_krnl_indexing.cpp [hunk 4]
```diff
@@ -26,12 +26,46 @@
...
 #include <dpnp_iface.hpp>
 #include "dpnp_fptr.hpp"
 #include "dpnpc_memory_adapter.hpp"
 #include "queue_sycl.hpp"
 
+template <typename _DataType1, typename _DataType2>
+class dpnp_choose_c_kernel;
+
+template <typename _DataType1, typename _DataType2>
+void dpnp_choose_c(void* result1, void* array1_in, std::vector<void*> choices1, size_t size)
```
Comment (shssf) [inline dpnp/backend/kernels/dpnp_krnl_indexing.cpp:40]: please do not use STL in the API. Change it as `void* choses, size_t choices_size`

**Comment by shssf** [comment]
Thank you. Many things are done right. Few of them needs to be addressed.

File: dpnp/dpnp_algo/dpnp_algo_indexing.pyx [hunk 5]
```diff
@@ -67,7 +67,9 @@ ctypedef void(*fptr_dpnp_nonzero_t)(const void * , void * , const size_t, const
 
 
 cpdef utils.dpnp_descriptor dpnp_choose(utils.dpnp_descriptor input, list choices1):
-    cdef shape_type_c input_shape = input.shape  
+    cdef shape_type_c input_shape = input.shape 
+    cdef shape_type_c choices_shape = (len(choices1),) 
+    cdef shape_type_c choice_shape = choices1[0].shape
```
Comment (densmirn) [inline dpnp/dpnp_algo/dpnp_algo_indexing.pyx:72]: Looks like we don't need the `shapes`, because we use only first element from them.

**Review by densmirn — COMMENTED** [review]
This PR shouldn't have code that doesn't related to the function `choose`.

**Review by shssf — APPROVED** [review]

**Pull request merged**